x-pack/libbeat/autodiscover/providers/aws/ec2/provider.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/elastic/beats/libbeat/logp"
+	awsauto "github.com/elastic/beats/x-pack/libbeat/autodiscover/providers/aws"
+	awscommon "github.com/elastic/beats/x-pack/libbeat/common/aws"
+)
+
+func init() {
+	autodiscover.Registry.AddProvider("aws_ec2", AutodiscoverBuilder)
+}
+
+// Provider implements autodiscover provider for aws EC2s.
+type Provider struct {
+	config        *awsauto.Config
+	bus           bus.Bus
+	templates     *template.Mapper
+	startListener bus.Listener
+	stopListener  bus.Listener
+	watcher       *watcher
+	uuid          uuid.UUID
+}
+
+// AutodiscoverBuilder is the main builder for this provider.
+func AutodiscoverBuilder(bus bus.Bus, uuid uuid.UUID, c *common.Config) (autodiscover.Provider, error) {
+	cfgwarn.Experimental("aws_ec2 autodiscover is experimental")
+
+	config := awsauto.DefaultConfig()
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	awsCfg, err := awscommon.GetAWSCredentials(
+		awscommon.ConfigAWS{
+			AccessKeyID:     config.AWSConfig.AccessKeyID,
+			SecretAccessKey: config.AWSConfig.SecretAccessKey,
+			SessionToken:    config.AWSConfig.SessionToken,
+			ProfileName:     config.AWSConfig.ProfileName,
+		})
+
+	// Construct MetricSet with a full regions list if there is no region specified.
+	if config.Regions == nil {
+		// set default region to make initial aws api call
+		awsCfg.Region = "us-west-1"
+		svcEC2 := ec2.New(awsCfg)
+		completeRegionsList, err := awsauto.GetRegions(svcEC2)
+		if err != nil {
+			return nil, err
+		}
+
+		config.Regions = completeRegionsList
+	}
+
+	var clients []ec2iface.ClientAPI
+	for _, region := range config.Regions {
+		if err != nil {
+			logp.Error(errors.Wrap(err, "error loading AWS config for aws_ec2 autodiscover provider"))
+		}
+		awsCfg.Region = region
+		clients = append(clients, ec2.New(awsCfg))
+	}
+
+	return internalBuilder(uuid, bus, config, newAPIFetcher(clients))
+}
+
+// internalBuilder is mainly intended for testing via mocks and stubs.
+// it can be configured to use a fetcher that doesn't actually hit the AWS API.
+func internalBuilder(uuid uuid.UUID, bus bus.Bus, config *awsauto.Config, fetcher fetcher) (*Provider, error) {
+	mapper, err := template.NewConfigMapper(config.Templates)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Provider{
+		config:    config,
+		bus:       bus,
+		templates: &mapper,
+		uuid:      uuid,
+	}
+
+	p.watcher = newWatcher(
+		fetcher,
+		config.Period,
+		p.onWatcherStart,
+		p.onWatcherStop,
+	)
+
+	return p, nil
+}
+
+// Start the autodiscover process.
+func (p *Provider) Start() {
+	p.watcher.start()
+}
+
+// Stop the autodiscover process.
+func (p *Provider) Stop() {
+	p.watcher.stop()
+}
+
+func (p *Provider) onWatcherStart(instanceID string, instance *ec2Instance) {
+	e := bus.Event{
+		"start":    true,
+		"provider": p.uuid,
+		"id":       instanceID,
+		"meta": common.MapStr{
+			"aws": common.MapStr{
+				"ec2":   instance.toMap(),
+			},
+			"cloud": instance.toCloudMap(),
```

Comment:
here meta represents what will end up as metadasta in the output of modules that get launched. We also want this  to be available to autodiscover templates. To do so you need to add it to the root of the event too:

{
..
"aws":  common.MapStr{
    "ec2":   instance.toMap(),
  }
  "cloud" ...
}
"meta":  common.MapStr{
  "aws": common.MapStr{
    "ec2":   instance.toMap(),
  }
  "cloud" ...
}

Reply:
Good to know, thanks! I added the data to autodiscover template so now in autodiscover config, for example in the equal condition, we don't need `meta.aws.ec2.tags.created-by`. `aws.ec2.tags.created-by: "ks"` should work. Same as hosts field in mysql config, `${data.aws.ec2.public.ip}` should work.